### DISCUSSION TITLE
[学习提问] 你好: ATOM_CAS 实现为什么是 atomic_compare_exchange_weak 而不是 atomic_compare_exchange_strong

### DISCUSSION THREAD

**Discussion by wangzhione** [cloudwu/skynet#1532] [General]
#define ATOM_CAS(ptr, oval, nval) atomic_compare_exchange_weak(ptr, &(oval), nval)

看文档说:
weak 版本的 CAS 允许偶然出乎意料的返回(比如在字段值和期待值一样的时候却返回了false), 
不过在一些循环算法中, 这是可以接受的. 通常它比起 strong 版本有更高的性能.

咱们这里用 atomic_compare_exchange_strong 是否更合适点呢?

**Comment by cloudwu** [comment]
既然没有 bug , 为什么要换成 strong .

**Comment by wangzhione** [comment]
有 外围 for 业务兜底, 业务上面没有 bug. 

观察到有些平台 atomic_compare_exchange_weak 和 atomic_compare_exchange_strong 底层实现一样的, 理论上也不会有 bug.

例如这样

`#define atomic_compare_exchange_strong_explicit(PTR, VAL, DES, SUC, FAIL) \
  __extension__								\
  ({									\
    __auto_type __atomic_compare_exchange_ptr = (PTR);			\
    __typeof__ (*__atomic_compare_exchange_ptr) __atomic_compare_exchange_tmp \
      = (DES);								\
    __atomic_compare_exchange (__atomic_compare_exchange_ptr, (VAL),	\
			       &__atomic_compare_exchange_tmp, 0,	\
			       (SUC), (FAIL));				\
  })

#define atomic_compare_exchange_strong(PTR, VAL, DES) 			   \
  atomic_compare_exchange_strong_explicit (PTR, VAL, DES, __ATOMIC_SEQ_CST, \
					   __ATOMIC_SEQ_CST)

#define atomic_compare_exchange_weak_explicit(PTR, VAL, DES, SUC, FAIL) \
  __extension__								\
  ({									\
    __auto_type __atomic_compare_exchange_ptr = (PTR);			\
    __typeof__ (*__atomic_compare_exchange_ptr) __atomic_compare_exchange_tmp \
      = (DES);								\
    __atomic_compare_exchange (__atomic_compare_exchange_ptr, (VAL),	\
			       &__atomic_compare_exchange_tmp, 1,	\
			       (SUC), (FAIL));				\
  })

#define atomic_compare_exchange_weak(PTR, VAL, DES)			\
  atomic_compare_exchange_weak_explicit (PTR, VAL, DES, __ATOMIC_SEQ_CST, \
					 __ATOMIC_SEQ_CST)`

区别在于 __atomic_compare_exchange 的 第 4 个参数上面, 用 1 标识 weak 调用, 0 默认 strong 调用. 
例如 gcc 一种实现 
https://code.woboq.org/gcc/libgcc/config/tilepro/atomic.c.html#164

`
#define __atomic_exchange_methods(type, size)                                \
--
bool                                                                        \
__atomic_compare_exchange_##size(volatile type* ptr, type* oldvalp,        \
type newval, bool weak __unused,        \
int models, int modelf __unused)        \
`

weak 是 unused 未被使用的.

**Comment by Cheungwq** [comment]
您好，您发给我的邮件已收到。

**Comment by cloudwu** [comment]
我的意思是，如果 skynet 在使用 CAS 的时候，都是基于 weak 的定义把事情做对。那么什么时候你想换成 strong 都没有关系。比如有的平台没有 weak 的实现，你可以转到 strong 使用。

反过来就不成立了。一旦用了 strong 语义，就再不能轻易的换回 weak 。

**Comment by wangzhione** [comment]
恩恩 明白更加清晰了. 谢谢你的指导

**Comment by fangcun010** [comment]
这里有一份关于哪些情况更应该使用weak还是strong的资料：

[atomic<> Weapons: The C++ Memory Model and Modern Hardware](https://onedrive.live.com/view.aspx?resid=4E86B0CF20EF15AD!24884&authkey=!AMtj_EflYn2507c)
![vs](https://user-images.githubusercontent.com/8509053/235309262-2d5294d2-8f18-4910-a2a1-f49c2a04b902.png)


里面提到，一般情况下weak用在循环CAS中(简单说就是while(CAS)空循环)，strong用在单次条件语句(区别于CAS空循环)中使用。

从这种提倡来看，用于CAS空循环的weak假设下次判断马上发生，那么这次结果就不太重要，会优先使用缓存作出**失败**判断，从而减少竞争(纯yy)。
